akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala

```diff
@@ -46,13 +47,45 @@ object ThreadPoolConfig {
   def reusableQueue(queue: BlockingQueue[Runnable]): QueueFactory = () ⇒ queue
 
   def reusableQueue(queueFactory: QueueFactory): QueueFactory = reusableQueue(queueFactory())
+
+  /**
+   * INTERNAL API
+   */
+  private[ThreadPoolConfig] final class BlockContextThreadFactory(
+    delegate: ThreadFactory, minUnblockedThreads: Int) extends AtomicInteger(0) with ThreadFactory {
+
+    @volatile var threadPoolExecutor: ThreadPoolExecutor = null
+    override def newThread(runnable: Runnable): Thread =
+      delegate.newThread(new Runnable with BlockContext {
+        override def run(): Unit = BlockContext.withBlockContext(this) {
+          incrementAndGet()
+          try runnable.run() finally { decrementAndGet() }
+        }
+
+        override def blockOn[T](thunk: ⇒ T)(implicit permission: CanAwait): T = {
+          @scala.annotation.tailrec def canBlock(): Boolean = {
+            val pre = get()
+            if (pre > 0 && pre > minUnblockedThreads) compareAndSet(pre, pre - 1) || canBlock()
```

Comment:
Also: no loop necessary if you just `decrementAndGet`, correcting if the result was negative.